<a href="https://colab.research.google.com/github/andazi/learning-pandas/blob/main/learning_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GETTING STARTED WITH PANDAS


FIRST WE IMPORT PANDAS AND NUMPY

In [1]:
import pandas as pd
import numpy as np

we then create a data objects using either the following

-    DataFrame
-    Series

DataFrame

A DataFrame comprises of group of series. It is a 2d array structure. It's like an excel spreadsheet in which the whole **Table** is the **DataFrame** while each columns are **series**

In [2]:
# DataFrame
pd.DataFrame({"Andazi" : [100, 65, 88, 90, 84], "Hermonie" : [99, 98, 100, 98, 100]})

Andazi  Hermonie
0     100        99
1      65        98
2      88       100
3      90        98
4      84       100

In [3]:
pd.DataFrame({"female names" : ["Olivia", "Emma", "Charlotte", "Amelia", "Sophia", "Ava", "Isabella", "Mia", "Evelyn", "Abigail"],
              "male names" : ["Liam", "Noah", "William", "James", "Oliver", "Benjamin", "Elijah", "Lucas", "Henry", "Alexander"]})


female names male names
0       Olivia       Liam
1         Emma       Noah
2    Charlotte    William
3       Amelia      James
4       Sophia     Oliver
5          Ava   Benjamin
6     Isabella     Elijah
7          Mia      Lucas
8       Evelyn      Henry
9      Abigail  Alexander

In [4]:
pd.DataFrame({"female names" : ["Olivia", "Emma", "Charlotte", "Amelia", "Sophia", "Ava", "Isabella", "Mia", "Evelyn", "Abigail"],
              "male names" : ["Liam", "Noah", "William", "James", "Oliver", "Benjamin", "Elijah", "Lucas", "Henry", "Alexander"]},
              index=["index 1", "index 2", "index 3", "index 4", "index 5", "index 6", "index 7", "index 8", "index 9", "index 10"])


female names male names
index 1        Olivia       Liam
index 2          Emma       Noah
index 3     Charlotte    William
index 4        Amelia      James
index 5        Sophia     Oliver
index 6           Ava   Benjamin
index 7      Isabella     Elijah
index 8           Mia      Lucas
index 9        Evelyn      Henry
index 10      Abigail  Alexander

Series

This is just like a list of items. Usually of the same datatype.


In [5]:
pd.Series([1,2,3,4,5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [6]:
pd.Series(["Apples", "Bananas", "Oranges", "Grapes", "Berries", "Melons", "Carrots", "Potatoes", "Lettuce", "Tomatoes", "Cucumbers", "Broccoli", "Bread", "Pasta", "Rice", "Cereal", "Milk", "Cheese", "Yogurt", "Butter", "Beef", "Chicken", "Pork", "Fish", "Eggs", "Nuts", "Seeds", "Sweets", "Drinks"])

0        Apples
1       Bananas
2       Oranges
3        Grapes
4       Berries
5        Melons
6       Carrots
7      Potatoes
8       Lettuce
9      Tomatoes
10    Cucumbers
11     Broccoli
12        Bread
13        Pasta
14         Rice
15       Cereal
16         Milk
17       Cheese
18       Yogurt
19       Butter
20         Beef
21      Chicken
22         Pork
23         Fish
24         Eggs
25         Nuts
26        Seeds
27       Sweets
28       Drinks
dtype: object

In [7]:
pd.Series(["Apples", "Bananas", "Oranges", "Grapes", "Berries", "Melons", "Carrots", "Potatoes", "Lettuce", "Tomatoes", "Cucumbers", "Broccoli", "Bread", "Pasta", "Rice", "Cereal", "Milk", "Cheese", "Yogurt", "Butter", "Beef", "Chicken", "Pork", "Fish", "Eggs", "Nuts", "Seeds", "Sweets", "Drinks"],
          index = ['product 1', 'product 2', 'product 3', 'product 4', 'product 5', 'product 6', 'product 7', 'product 8', 'product 9', 'product 10', 'product 11', 'product 12', 'product 13', 'product 14', 'product 15', 'product 16', 'product 17', 'product 18', 'product 19', 'product 20', 'product 21', 'product 22', 'product 23', 'product 24', 'product 25', 'product 26', 'product 27', 'product 28', 'product 29'],
          name = "Shopping List")

product 1        Apples
product 2       Bananas
product 3       Oranges
product 4        Grapes
product 5       Berries
product 6        Melons
product 7       Carrots
product 8      Potatoes
product 9       Lettuce
product 10     Tomatoes
product 11    Cucumbers
product 12     Broccoli
product 13        Bread
product 14        Pasta
product 15         Rice
product 16       Cereal
product 17         Milk
product 18       Cheese
product 19       Yogurt
product 20       Butter
product 21         Beef
product 22      Chicken
product 23         Pork
product 24         Fish
product 25         Eggs
product 26         Nuts
product 27        Seeds
product 28       Sweets
product 29       Drinks
Name: Shopping List, dtype: object

Reading CSV files

In [8]:
california_housing = pd.read_csv("/content/sample_data/california_housing_test.csv")

In [9]:
# check how large the resulting DateFrame
california_housing.shape

(3000, 9)

In [10]:
# displaying the first five rows to examine the data
california_housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   
1    -118.30     34.26                43.0       1510.0           310.0   
2    -117.81     33.78                27.0       3589.0           507.0   
3    -118.36     33.82                28.0         67.0            15.0   
4    -119.67     36.33                19.0       1241.0           244.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0  
1       809.0       277.0         3.5990            176500.0  
2      1484.0       495.0         5.7934            270500.0  
3        49.0        11.0         6.1359            330000.0  
4       850.0       237.0         2.9375             81700.0

Using `info` to get information about your dataset

In [66]:
california_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           3000 non-null   float64
 1   latitude            3000 non-null   float64
 2   housing_median_age  3000 non-null   float64
 3   total_rooms         3000 non-null   float64
 4   total_bedrooms      3000 non-null   float64
 5   population          3000 non-null   float64
 6   households          3000 non-null   float64
 7   median_income       3000 non-null   float64
 8   median_house_value  3000 non-null   float64
dtypes: float64(9)
memory usage: 211.1 KB


```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           3000 non-null   float64
 1   latitude            3000 non-null   float64
 2   housing_median_age  3000 non-null   float64
 3   total_rooms         3000 non-null   float64
 4   total_bedrooms      3000 non-null   float64
 5   population          3000 non-null   float64
 6   households          3000 non-null   float64
 7   median_income       3000 non-null   float64
 8   median_house_value  3000 non-null   float64
dtypes: float64(9)
memory usage: 211.1 KB
```

With the above data, we can understand that we have **9 columns** in which has an index of **3000**.

You can get the amount of data in each column and all required a value(non-null). **float64** means all data in those columns are integers.

This dataset memory usage is **211.1 KB**

In [11]:
students = pd.DataFrame({"female names" : ["Olivia", "Emma", "Charlotte", "Amelia", "Sophia", "Ava", "Isabella", "Mia", "Evelyn", "Abigail"],
              "male names" : ["Liam", "Noah", "William", "James", "Oliver", "Benjamin", "Elijah", "Lucas", "Henry", "Alexander"]},
              index=["index 1", "index 2", "index 3", "index 4", "index 5", "index 6", "index 7", "index 8", "index 9", "index 10"])


In [12]:
# saving it in csv format
students.to_csv("list_students_names.csv")

In [13]:
store_items = ["Apples", "Bananas", "Oranges", "Grapes", "Berries", "Melons", "Carrots", "Potatoes", "Lettuce", "Tomatoes", "Cucumbers", "Broccoli", "Bread", "Pasta", "Rice", "Cereal", "Milk", "Cheese", "Yogurt", "Butter", "Beef", "Chicken", "Pork", "Fish", "Eggs", "Nuts", "Seeds", "Sweets", "Drinks"]
index = ['product 1', 'product 2', 'product 3', 'product 4', 'product 5', 'product 6', 'product 7', 'product 8', 'product 9', 'product 10', 'product 11', 'product 12', 'product 13', 'product 14', 'product 15', 'product 16', 'product 17', 'product 18', 'product 19', 'product 20', 'product 21', 'product 22', 'product 23', 'product 24', 'product 25', 'product 26', 'product 27', 'product 28', 'product 29']


products = pd.Series(store_items, index, name = "Shopping List")
products.to_csv("Shopping_List.csv")

By default, when you read a csv file. a default index is inserted which starts from 0 but you can set it with the `index_col = 0`, this sets the very first column to be the beginning of the Table

In [14]:
reading_products = pd.read_csv("Shopping_List.csv")
reading_products

Unnamed: 0 Shopping List
0    product 1        Apples
1    product 2       Bananas
2    product 3       Oranges
3    product 4        Grapes
4    product 5       Berries
5    product 6        Melons
6    product 7       Carrots
7    product 8      Potatoes
8    product 9       Lettuce
9   product 10      Tomatoes
10  product 11     Cucumbers
11  product 12      Broccoli
12  product 13         Bread
13  product 14         Pasta
14  product 15          Rice
15  product 16        Cereal
16  product 17          Milk
17  product 18        Cheese
18  product 19        Yogurt
19  product 20        Butter
20  product 21          Beef
21  product 22       Chicken
22  product 23          Pork
23  product 24          Fish
24  product 25          Eggs
25  product 26          Nuts
26  product 27         Seeds
27  product 28        Sweets
28  product 29        Drinks

In [15]:
reading_index = pd.read_csv("Shopping_List.csv", index_col = 0)

In [16]:
reading_index

Shopping List
product 1         Apples
product 2        Bananas
product 3        Oranges
product 4         Grapes
product 5        Berries
product 6         Melons
product 7        Carrots
product 8       Potatoes
product 9        Lettuce
product 10      Tomatoes
product 11     Cucumbers
product 12      Broccoli
product 13         Bread
product 14         Pasta
product 15          Rice
product 16        Cereal
product 17          Milk
product 18        Cheese
product 19        Yogurt
product 20        Butter
product 21          Beef
product 22       Chicken
product 23          Pork
product 24          Fish
product 25          Eggs
product 26          Nuts
product 27         Seeds
product 28        Sweets
product 29        Drinks

Using the **dot notation** to access columns(fields)

```
data_name.column_name
```
If you observe, the output is exactly like in **Series**

In [17]:
california_housing.households

0        606.0
1        277.0
2        495.0
3         11.0
4        237.0
         ...  
2995     607.0
2996    1036.0
2997     220.0
2998      14.0
2999     260.0
Name: households, Length: 3000, dtype: float64

Using the **indexing operator in object** to access columns(fields)

```
data_name['column_name']
```
Using this method is better because some column names do contain whiteSpace, so we can't write `california.total rooms`. This would output an error

In [18]:
california_housing['population']

0       1537.0
1        809.0
2       1484.0
3         49.0
4        850.0
         ...  
2995    1258.0
2996    3496.0
2997     693.0
2998      46.0
2999     753.0
Name: population, Length: 3000, dtype: float64

In [19]:
# We can go further to access specific item in a list
california_housing['population'][2999]

753.0

In [20]:
# checking the dataType
type(california_housing)


pandas.core.frame.DataFrame

In [21]:
type(california_housing['population'])

pandas.core.series.Series

Indexing in Pandas is a little bit diffrent, we use `iloc` or `loc` operator.

In [22]:
california_housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   
1    -118.30     34.26                43.0       1510.0           310.0   
2    -117.81     33.78                27.0       3589.0           507.0   
3    -118.36     33.82                28.0         67.0            15.0   
4    -119.67     36.33                19.0       1241.0           244.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0  
1       809.0       277.0         3.5990            176500.0  
2      1484.0       495.0         5.7934            270500.0  
3        49.0        11.0         6.1359            330000.0  
4       850.0       237.0         2.9375             81700.0

Accessing rows, we can use the **index-based selection** or the **label-based selection**

The index-based selection usings integers like `california_housing.iloc[4]`

The code above says, in *california_housing* and *return the whole row in the 4th index*

In [23]:
# checking the row(period) with an index of 4
california_housing.iloc[4]

longitude              -119.6700
latitude                 36.3300
housing_median_age       19.0000
total_rooms            1241.0000
total_bedrooms          244.0000
population              850.0000
households              237.0000
median_income             2.9375
median_house_value    81700.0000
Name: 4, dtype: float64

In [24]:
carlifornia_housing_9 = pd.DataFrame(california_housing.iloc[9])
carlifornia_housing_9

9
longitude             -118.0200
latitude                34.0800
housing_median_age      31.0000
total_rooms           2402.0000
total_bedrooms         632.0000
population            2830.0000
households             603.0000
median_income            2.3333
median_house_value  164200.0000

In [25]:
# checking the row(period) with an index of 9
california_housing.iloc[0]

longitude               -122.0500
latitude                  37.3700
housing_median_age        27.0000
total_rooms             3885.0000
total_bedrooms           661.0000
population              1537.0000
households               606.0000
median_income              6.6085
median_house_value    344700.0000
Name: 0, dtype: float64

In [26]:
# checking the last 5 rows(periods)
california_housing.iloc[-5:]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2995    -119.86     34.42                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0

In [27]:
# accessing specific rows
california_housing.iloc[[0,10,100,1000]]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       -122.05     37.37                27.0       3885.0           661.0   
10      -118.24     33.98                45.0        972.0           249.0   
100     -118.30     34.02                49.0       2120.0           483.0   
1000    -119.75     34.40                31.0       1997.0           299.0   

      population  households  median_income  median_house_value  
0         1537.0       606.0         6.6085            344700.0  
10        1288.0       261.0         2.2054            125000.0  
100       1522.0       416.0         1.8500            116800.0  
1000       826.0       301.0         6.8927            500001.0

You can select the exact column to view, using

```
[:, column_index]
```

In [28]:
# the below code check for the 3 index(total_rooms) of the last 5 rows
california_housing.iloc[-5:,3]

2995    1450.0
2996    5257.0
2997     956.0
2998      96.0
2999    1765.0
Name: total_rooms, dtype: float64

In [29]:
# accessing specific rows and column
california_housing.iloc[[0,10,1000], 2]

0       27.0
10      45.0
1000    31.0
Name: housing_median_age, dtype: float64

The label-based selection usings integers like `california_housing.loc[4, 'population']`

The code above says, in *california_housing* and *return the value of population 4th index*

`loc` returns the value not the whole rows

In [30]:
california_housing.loc[4, 'population']

850.0

In [31]:
california_housing.loc[4, 'longitude']

-119.67

In [32]:
california_housing.loc[4, 'latitude']

36.33

In [33]:
california_housing.loc[4, 'housing_median_age']

19.0

In [34]:
# you can return all at once based on needs
california_housing.loc[4, ['population', 'longitude', 'latitude', 'housing_median_age']]

population            850.00
longitude            -119.67
latitude               36.33
housing_median_age     19.00
Name: 4, dtype: float64

In [35]:
# you can return all at once based on needs
california_housing.loc[[4, 40, 400], ['population', 'longitude', 'latitude', 'housing_median_age']]

population  longitude  latitude  housing_median_age
4         850.0    -119.67     36.33                19.0
40        961.0    -119.32     36.20                15.0
400      1078.0    -121.72     36.81                18.0

**Combining Series to DataFrame**

Lets add two or more Series to form a DataFrame and we would be using the following methods


- concat method
- key and values method

- **concat**
The `concat` method is versatile and easy to use as we just concat two or more Series together.
we use the `concat()` method, and `axis=1` as argument.

```
first_name = pd.Series(['James', 'Luna', 'Cleopatra'])
last_name = pd.Series(['Potter', 'Lovegood', 'Achilles'])

df = pd.concat([last_name, first_name], axis=1)
```

In [36]:
first_name = pd.Series(['James', 'Luna', 'Cleopatra'])
last_name = pd.Series(['Potter', 'Lovegood', 'Achilles'])

df = pd.concat([last_name, first_name], axis=1)
df

0          1
0    Potter      James
1  Lovegood       Luna
2  Achilles  Cleopatra

In [37]:
id_series = pd.Series([1, 2, 3, 4, 5], name = 'ID')
fullnames_series = pd.Series(["John Doe", "Jane Doe", "Mary Smith", "Peter Jones", "Susan Williams"], name= 'Fullname')
age_series = pd.Series([25, 30, 35, 40, 45], name = "Age")
rank_series = pd.Series(["Junior", "Senior", "Manager", "Director", "Executive"], name = 'Position')

employees = pd.concat([id_series, fullnames_series, age_series, rank_series], axis=1)
employees

ID        Fullname  Age   Position
0   1        John Doe   25     Junior
1   2        Jane Doe   30     Senior
2   3      Mary Smith   35    Manager
3   4     Peter Jones   40   Director
4   5  Susan Williams   45  Executive

- key value method

This method is quite simple too. You only need to create a dictionary containing `{keys : values}`

**the keys being the column names**

**the value being the Series**

Then you call your DataFrame as usual

```
first_name = pd.Series(['James', 'Luna', 'Cleopatra'])
last_name = pd.Series(['Potter', 'Lovegood', 'Achilles'])

cast = {'firstname' : first_name, 'lastname' : last_name}

cast = pd.DataFrame(cast)

```

In [38]:
first_name = pd.Series(['James', 'Luna', 'Cleopatra'])
last_name = pd.Series(['Potter', 'Lovegood', 'Achilles'])

cast = {'firstname' : first_name, 'lastname' : last_name}

cast = pd.DataFrame(cast)
cast

firstname  lastname
0      James    Potter
1       Luna  Lovegood
2  Cleopatra  Achilles

**Manupulating the index**

When we read a csv file, an index is created counting from **0** to the **nth** row.

we can use the `set_index()` to create or set an index.

In [39]:
employees.head()
# an index is inserted by default but we do not want it

ID        Fullname  Age   Position
0   1        John Doe   25     Junior
1   2        Jane Doe   30     Senior
2   3      Mary Smith   35    Manager
3   4     Peter Jones   40   Director
4   5  Susan Williams   45  Executive

In [40]:
employees = employees.set_index('ID')
employees

Fullname  Age   Position
ID                                
1         John Doe   25     Junior
2         Jane Doe   30     Senior
3       Mary Smith   35    Manager
4      Peter Jones   40   Director
5   Susan Williams   45  Executive

In [41]:
import random

# Create a list of data for the DataFrame
data = []
for i in range(50):
    id = i + 1
    name = random.choice(["Amina", "Abubakar", "Ama", "Ayo", "Blessing", "Chinedu", "David", "Ezekiel", "Femi", "Gladys", "Hamza", "Ibrahim", "Jacqueline", "Kampala", "Lily", "Maryam", "Naomi", "Oluwatobi", "Peter", "Rita", "Samuel", "Tunde", "Umar", "Victor", "Wanjiru", "Xhosa", "Yusuf", "Zara"])
    age = i % 10 + 20
    rank = ["Junior", "Senior", "Manager", "Director", "Executive"][i % 5]
    email = f"{name}.{random.randint(100, 999)}@example.com"
    data.append({"id": id, "name": name, "age": age, "rank": rank, "email": email})

# Create the DataFrame
job_applicants = pd.DataFrame(data)

job_applicants.to_csv('job_applicants.csv')




In [42]:
job_applicants.set_index('id').head()

name  age       rank                    email
id                                                  
1      Rita   20     Junior     Rita.721@example.com
2       Ayo   21     Senior      Ayo.893@example.com
3    Maryam   22    Manager   Maryam.685@example.com
4     Hamza   23   Director    Hamza.606@example.com
5   Ibrahim   24  Executive  Ibrahim.352@example.com

In [43]:

# Create a list of product IDs
product_ids = [100, 101, 102, 103, 104, 105, 106, 107, 108, 109]

# Create a list of products
products = ["iPhone", "iPad", "MacBook Pro", "Apple Watch", "AirPods", "Apple TV", "HomePod", "Beats Headphones", "Apple Pencil", "Magic Mouse"]

# Create a list of quantities
quantities = [1, 2, 11, 4, 5, 6, 7, 8, 9, 10]

# Create a list of prices
prices = [1000, 800, 1500, 300, 200, 200, 300, 250, 100, 100]

# Create a DataFrame
products_table = pd.DataFrame({"product_Id": product_ids, "product": products, "quantity": quantities, "price": prices})



In [44]:
products_table.set_index('product_Id')

product  quantity  price
product_Id                                   
100                   iPhone         1   1000
101                     iPad         2    800
102              MacBook Pro        11   1500
103              Apple Watch         4    300
104                  AirPods         5    200
105                 Apple TV         6    200
106                  HomePod         7    300
107         Beats Headphones         8    250
108             Apple Pencil         9    100
109              Magic Mouse        10    100

We can now select

Why don't we try **conditional selection** Instead of just selecting, lets select based on a condition

```
products_table.price == 1000
```
returns `True` or `False` based on the condition

In [45]:
# checking rows whose price is exactly 1000
products_table.price == 1000

0     True
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: price, dtype: bool

To return the values that satisfies the condition

we use the `label based selection` `loc`

```
products_table.loc[products_table.price == 1000]
```

In [46]:
# return rows whose price is exactly 1000

products_table.loc[products_table.price == 1000]

product_Id product  quantity  price
0         100  iPhone         1   1000

In [47]:
# return rows whose prices exceed 500

products_table.loc[products_table.price > 500]

product_Id      product  quantity  price
0         100       iPhone         1   1000
1         101         iPad         2    800
2         102  MacBook Pro        11   1500

In [48]:
# return rows whose prices exceed 500 and quantity is less than 3

products_table.loc[(products_table.price > 500) & (products_table.quantity <= 2)]

product_Id product  quantity  price
0         100  iPhone         1   1000
1         101    iPad         2    800

Assigning data / creating new column

You can assign with a constant data

```
table_name['new_column_name'] = 'data'
```

or with a logical way

```
table_name['new_column_name'] = data1 ** 3

```

In [49]:
products_table['total'] = products_table['quantity'] * products_table['price']

In [50]:
products_table.head()

product_Id      product  quantity  price  total
0         100       iPhone         1   1000   1000
1         101         iPad         2    800   1600
2         102  MacBook Pro        11   1500  16500
3         103  Apple Watch         4    300   1200
4         104      AirPods         5    200   1000

Extracting Insight from your data

using **summary Function** and **Maps**

In the summary funtions, We use `describe()` which returns high summary of the selected column.

The output of the `describe()` function will be a table that shows the following statistics for each column:

- Count: The number of non-NA/null observations.
- Mean: The mean of the values.
- Std: The standard deviation of the observations.
- Min: The minimum of the values.
- 25%: The 25th percentile of the values.
- 50%: The 50th percentile of the values, also known as the median.
- 75%: The 75th percentile of the values.
- Max: The maximum of the values.

```
table_name.column_name.describe()
```

In [51]:
# describe()
products_table.total.describe()

count       10.000000
mean      2850.000000
std       4815.311228
min        900.000000
25%       1000.000000
50%       1200.000000
75%       1900.000000
max      16500.000000
Name: total, dtype: float64

We cango as far as to extract specific data.

```
table_name.column_name.count()
```

In [52]:
products_table.quantity.count()

10

`idxmax()` is use to check the index of the maximum value and takes none or one argument

In [53]:
# the result shows the index of the maximum value(price) in a table
products_table.price.idxmax()

2

In [54]:
# returning the Series for the maximum value
products_table.loc[products_table.price.idxmax()]

product_Id            102
product       MacBook Pro
quantity               11
price                1500
total               16500
Name: 2, dtype: object

In [55]:
products_table.groupby(['quantity', 'price']).apply(lambda df : df.loc[df.quantity.idxmax()])

product_Id           product  quantity  price  total
quantity price                                                      
1        1000          100            iPhone         1   1000   1000
2        800           101              iPad         2    800   1600
4        300           103       Apple Watch         4    300   1200
5        200           104           AirPods         5    200   1000
6        200           105          Apple TV         6    200   1200
7        300           106           HomePod         7    300   2100
8        250           107  Beats Headphones         8    250   2000
9        100           108      Apple Pencil         9    100    900
10       100           109       Magic Mouse        10    100   1000
11       1500          102       MacBook Pro        11   1500  16500

**Maps**

Just like the word 'map', we are linking one value to another.

We usually create new data from existing ones to work with in format we need.

we can either use the `map()` or `apply()` method

**map()** returns a new Series where all the values selected have been transformed by your function

```
table_mean = table_name.column_name.mean()
table_name.column_name.map(lambda x : x - table_mean)
```

In [56]:
# getting the mean
products_mean = products_table.price.mean()

# using lamba to get the mean deviation
# this is called 'centering' transformation is a common preprocessing step before applying various machine learning algorithms.
products_table.price.map(lambda m : m - products_mean)

0     525.0
1     325.0
2    1025.0
3    -175.0
4    -275.0
5    -275.0
6    -175.0
7    -225.0
8    -375.0
9    -375.0
Name: price, dtype: float64

In [57]:
# editing the product ID with initials '0x34'
products_table.product_Id.map(lambda x : '0x34' + str(x))

0    0x34100
1    0x34101
2    0x34102
3    0x34103
4    0x34104
5    0x34105
6    0x34106
7    0x34107
8    0x34108
9    0x34109
Name: product_Id, dtype: object

**apply()**

This is use when we want to apply a custom functions to each rows

```
table_mean = table_name.column_name.mean()

def custom_fn(row):
    # mean deviation
    row.points = row.points - table_mean
    return row

column_name.apply(custom_fn, axis=column)
```

In [58]:
# getting the mean
products_mean = products_table.price.mean()

#function that returns the mean deviation

def products_mean_deviation(table):
    table.price = table.price - products_mean

    return table

In [59]:
# calling the apply() method
products_table.apply(products_mean_deviation, axis='columns')

product_Id           product  quantity   price  total
0         100            iPhone         1   525.0   1000
1         101              iPad         2   325.0   1600
2         102       MacBook Pro        11  1025.0  16500
3         103       Apple Watch         4  -175.0   1200
4         104           AirPods         5  -275.0   1000
5         105          Apple TV         6  -275.0   1200
6         106           HomePod         7  -175.0   2100
7         107  Beats Headphones         8  -225.0   2000
8         108      Apple Pencil         9  -375.0    900
9         109       Magic Mouse        10  -375.0   1000

In [60]:
def product_unique_id(table):
    # new product id for new brand starts with '0x42'
    table.product_Id = '0x42' + str(table.product_Id)

    return table

In [61]:
# using apply()
products_table.apply(product_unique_id, axis='columns')

product_Id           product  quantity  price  total
0    0x42100            iPhone         1   1000   1000
1    0x42101              iPad         2    800   1600
2    0x42102       MacBook Pro        11   1500  16500
3    0x42103       Apple Watch         4    300   1200
4    0x42104           AirPods         5    200   1000
5    0x42105          Apple TV         6    200   1200
6    0x42106           HomePod         7    300   2100
7    0x42107  Beats Headphones         8    250   2000
8    0x42108      Apple Pencil         9    100    900
9    0x42109       Magic Mouse        10    100   1000

**Grouping and Sorting**



**Groupwise analysis**

In [62]:
# creates a Series of sorted quantity and attached their product
products_table.groupby('quantity').product.min()

quantity
1               iPhone
2                 iPad
4          Apple Watch
5              AirPods
6             Apple TV
7              HomePod
8     Beats Headphones
9         Apple Pencil
10         Magic Mouse
11         MacBook Pro
Name: product, dtype: object

We can go as far as to manipulate our dataset.

Let try and use `apply()`

In [63]:
# lets check for price greater than 900
products_table.groupby('price').apply(lambda df : df.price >= 1000)

price   
100    8    False
       9    False
200    4    False
       5    False
250    7    False
300    3    False
       6    False
800    1    False
1000   0     True
1500   2     True
Name: price, dtype: bool

In [64]:
# lets check for price greater than 900
products_table.groupby('price').apply(lambda df : df.price.iloc[0] > 500)

price
100     False
200     False
250     False
300     False
800      True
1000     True
1500     True
dtype: bool

In [65]:
products_table.iloc[0]

product_Id       100
product       iPhone
quantity           1
price           1000
total           1000
Name: 0, dtype: object